yb-voyager/src/tgtdb/oracle.go

```diff
@@ -0,0 +1,478 @@
+            EXECUTE IMMEDIATE 'CREATE USER %s IDENTIFIED BY "password"';
+        END IF;
+    END;
+END;`, BATCH_METADATA_TABLE_SCHEMA, BATCH_METADATA_TABLE_SCHEMA)
+	grantQuery := fmt.Sprintf(`GRANT CONNECT, RESOURCE TO %s`, BATCH_METADATA_TABLE_SCHEMA)
+	alterQuery := fmt.Sprintf(`ALTER USER %s QUOTA UNLIMITED ON USERS`, BATCH_METADATA_TABLE_SCHEMA)
+	createTableQuery := fmt.Sprintf(`BEGIN
+		EXECUTE IMMEDIATE 'CREATE TABLE %s (
+			data_file_name VARCHAR2(250),
+			batch_number NUMBER(10),
+			schema_name VARCHAR2(250),
+			table_name VARCHAR2(250),
+			rows_imported NUMBER(19),
+			PRIMARY KEY (data_file_name, batch_number, schema_name, table_name)
+		)';
+	EXCEPTION
+		WHEN OTHERS THEN
+			IF SQLCODE != -955 THEN
+				RAISE;
+			END IF;
+	END;`, BATCH_METADATA_TABLE_NAME)
+
+	cmds := []string{
+		createUserQuery,
+		grantQuery,
+		alterQuery,
+		createTableQuery,
+	}
+
+	maxAttempts := 12
+	var err error
+
+outer:
+	for _, cmd := range cmds {
+		for attempt := 1; attempt <= maxAttempts; attempt++ {
+			log.Infof("Executing on target: [%s]", cmd)
+			conn := db.GetConnection()
+			_, err = conn.ExecContext(context.Background(), cmd)
+			if err == nil {
+				// No error. Move on to the next command.
+				continue outer
+			}
+			log.Warnf("Error while running [%s] attempt %d: %s", cmd, attempt, err)
+			time.Sleep(5 * time.Second)
+			err2 := db.reconnect()
+			if err2 != nil {
+				break
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("create ybvoyager schema on target: %w", err)
+		}
+	}
+	return nil
+}
+
+func (db *TargetOracleDB) GetNonEmptyTables(tables []string) []string {
+	result := []string{}
+
+	for _, table := range tables {
+		log.Infof("Checking if table %s.%s is empty", db.tconf.Schema, table)
+		tmp := false
+		stmt := fmt.Sprintf("SELECT 1 FROM %s.%s WHERE ROWNUM <= 1", db.tconf.Schema, table)
+		err := db.conn_.QueryRow(stmt).Scan(&tmp)
+		if err != nil {
+			log.Errorf("Failed to check if table %s is empty: %v", table, err)
+			continue
+		}
+		if tmp {
+			result = append(result, table)
+		}
+	}
+
+	return result
+}
+
+func (db *TargetOracleDB) IsNonRetryableCopyError(err error) bool {
+	return false
+}
+
+func (db *TargetOracleDB) ImportBatch(batch Batch, args *ImportBatchArgs, exportDir string) (int64, error) {
+	var rowsAffected int64
+	var err error
+	copyFn := func(conn *sql.Conn) (bool, error) {
+		rowsAffected, err = db.importBatch(conn, batch, args, exportDir)
+		return false, err
+	}
+	err = db.WithConn(copyFn)
+	return rowsAffected, err
+}
+
+func (db *TargetOracleDB) WithConn(fn func(*sql.Conn) (bool, error)) error {
+	var err error
+	var conn *sql.Conn
+	if db.conn_ == nil {
+		err = db.reconnect()
+		if err != nil {
+			return err
+		}
+	}
+	retry := true
+	for retry {
+		conn, err = db.conn_.Conn(context.Background())
+		if err != nil {
+			return fmt.Errorf("get connection from target db: %w", err)
+		}
+		retry, err = fn(conn)
+		if err != nil {
+			// On err, drop the connection.
+			conn.Close()
```

Comment:
```
// Conn returns a single connection by either opening a new connection
// or returning an existing connection from the connection pool. Conn will
// block until either a connection is returned or ctx is canceled.
// Queries run on the same Conn will be run in the same database session.
//
// Every Conn must be returned to the database pool after use by
// calling Conn.Close.
```
If `err` is `nil`, you're not returning the `conn` back to the pool. This way the pool will soon run out of connections.

Consider using the `*sql.Conn` object i.e. `tdb.conn` (as per previously given comments). Then all you need is to call `reconnect()` before retrying.